### PR TITLE
Add client configurations

### PR DESCRIPTION
**Pull request by ynguyensegment**
Add new function to allow creating clients with configurations

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -48,17 +51,45 @@ type Client struct {
 }
 
 func New(writeKey string) *Client {
+	return NewWithConfig(writeKey, Config{})
+}
+
+func NewWithConfig(writeKey string, config Config) *Client {
```
Comment (dk1027) [inline client.go:64]: What do you think if `Config` contains the writeKey?

```
func New(writeKey string) *Client {
  return NewWithConfig(Config{WriteKey: writeKey})
}

func NewWithConfig(config Config) *Client {
...
}
```

File: client.go [hunk 2]
```diff
@@ -48,17 +51,45 @@ type Client struct {
...
-		MaxBatchInterval: 10 * time.Second,
-		semaphore:        make(semaphore.Semaphore, 10),
+		Config:    conf,
+		writeKey:  writeKey,
+		cmap:      newConcurrentMap(),
+		semaphore: make(semaphore.Semaphore, 10),
+	}
+}
+
+func getFinalConfig(c Config) Config {
```
Comment (dk1027) [inline client.go]: `getFinalConfig` doesn't really explain much.
Could you consider rename/refactor to 
```
func applyDefaultConfig(c *Config){
// sets default inline
}
```

File: client.go (hunk 1 above)
Comment (ynguyensegment) [reply client.go:64]: Configs tend to imply optionality. The writekey is not optional. I am not sure it belongs in the config

File: client.go (hunk 2 above)
Comment (ynguyensegment) [reply client.go]: naming is hard. How do parents do this

**Review by tysonmote — APPROVED** [review]

**Comment by tysonmote** [comment]
I'm going to merge and tag this release as v2.0.0 because it's a breaking API change (albeit minor).

```
▶ go test ./... -v
=== RUN   TestBuffer
=== RUN   TestAddMultiple
--- PASS: TestAddMultiple (0.00s)
=== RUN   TestAddMultipleMarshalReset
--- PASS: TestAddMultipleMarshalReset (0.00s)
=== RUN   TestAddMultipleReset
--- PASS: TestAddMultipleReset (0.00s)
=== RUN   TestAddNew
--- PASS: TestAddNew (0.00s)
=== RUN   TestMarshalEmptyArray
--- PASS: TestMarshalEmptyArray (0.00s)
=== RUN   TestMarshalSingleArray
--- PASS: TestMarshalSingleArray (0.00s)
=== RUN   TestNewBuffer
--- PASS: TestNewBuffer (0.00s)
--- PASS: TestBuffer (0.00s)
=== RUN   TestClient
=== RUN   TestChannelFlow
--- PASS: TestChannelFlow (0.25s)
=== RUN   TestClose
--- PASS: TestClose (0.00s)
=== RUN   TestNewClient
--- PASS: TestNewClient (0.00s)
=== RUN   TestSetErrors
--- PASS: TestSetErrors (0.00s)
=== RUN   TestSetFull
--- PASS: TestSetFull (0.00s)
=== RUN   TestSetOnce
--- PASS: TestSetOnce (0.00s)
--- PASS: TestClient (0.25s)
PASS
ok  	github.com/segmentio/objects-go	0.614s
```

**Pull request merged**